File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        result = []
        nums.sort()
        for i in range(len(nums) - 2):
            l = i + 1
            r = len(nums) - 1
            if i > 0 and nums[i] == nums[i -1]:
                continue
            target = 0 - nums[i]
            while (l < r):
                curr = nums[l] + nums[r]
                if curr == target:
                    result.append([nums[i], nums[l], nums[r]])
                    while (l < r and nums[l] == nums[l + 1]):
                        l += 1
                    
                    while (l < r and nums[r] == nums[r - 1]):
                        r -= 1
                    l += 1
                    r -= 1
                elif curr < target:
                    l += 1
                else:
                    r -= 1
        return result

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> list = new ArrayList();
        backtrack(list, new ArrayList<>(), target, 0, nums);
        return list;
    }

    public void backtrack(List<List<Integer>> list, List<Integer> curr, int target, int temp, int[] nums) {
        if (target == 0) {
            list.add(new ArrayList<>(curr));
        }
        for (int i = temp; i < nums.length; i++) {
            if (target - nums[i] >= 0) {
                curr.add(nums[i]);
                backtrack(list, curr, target - nums[i], i, nums);
                curr.remove(curr.size() - 1);
            }
        }
    }
}
